data_structures/src/chain.rs

```diff
@@ -3344,6 +3370,119 @@ mod tests {
+        let dr_pointer2 = drt2.hash();
+
+        let dro3 = DataRequestOutput {
+            witnesses: 1,
+            commit_fee: 500,
+            ..Default::default()
+        };
+        let drb3 = DRTransactionBody::new(vec![], vec![], dro3);
+        let drt3 = DRTransaction::new(
+            drb3,
+            vec![KeyedSignature {
+                signature: Default::default(),
+                public_key,
+            }],
+        );
+        let dr_pointer3 = drt3.hash();
+
+        let mut dr_pool = DataRequestPool::default();
+        dr_pool
+            .add_data_request(0, drt1, &Default::default())
+            .unwrap();
+        dr_pool
+            .add_data_request(0, drt2, &Default::default())
+            .unwrap();
+        dr_pool
+            .add_data_request(0, drt3, &Default::default())
+            .unwrap();
+
+        let c1 = Hash::SHA256([1; 32]);
+        let mut cb1 =
+            CommitTransactionBody::without_collateral(dr_pointer1, c1, Default::default());
+        cb1.collateral = vec![Input::default()];
+        let t1 = Transaction::Commit(CommitTransaction {
+            body: cb1,
+            signatures: vec![KeyedSignature::default()],
+        });
+
+        let c2 = Hash::SHA256([2; 32]);
+        let mut cb2 =
+            CommitTransactionBody::without_collateral(dr_pointer2, c2, Default::default());
+        cb2.collateral = vec![Input {
+            output_pointer: OutputPointer {
+                transaction_id: Default::default(),
+                output_index: 2,
+            },
+        }];
+        let t2 = Transaction::Commit(CommitTransaction {
+            body: cb2,
+            signatures: vec![KeyedSignature::default()],
+        });
+
+        // Commitment with same Input
+        let c3 = Hash::SHA256([3; 32]);
+        let mut cb3 =
+            CommitTransactionBody::without_collateral(dr_pointer3, c3, Default::default());
+        cb3.collateral = vec![Input::default()];
+        let t3 = Transaction::Commit(CommitTransaction {
+            body: cb3,
+            signatures: vec![KeyedSignature::default()],
+        });
+
+        let mut transactions_pool = TransactionsPool::default();
+        transactions_pool.insert(t1);
+        assert_eq!(transactions_pool.co_transactions.len(), 1);
+        assert_eq!(transactions_pool.co_hash_index.len(), 1);
+        transactions_pool.insert(t2);
+        assert_eq!(transactions_pool.co_transactions.len(), 2);
+        assert_eq!(transactions_pool.co_hash_index.len(), 2);
+        transactions_pool.insert(t3);
+        assert_eq!(transactions_pool.co_transactions.len(), 3);
+        assert_eq!(transactions_pool.co_hash_index.len(), 3);
+
+        let (commits_vec, commit_fees) = transactions_pool.remove_commits(&dr_pool);
+        assert_eq!(commits_vec.len(), 2);
+        // If the commit transactions with the same input would be returned,
+        // the commit fees exceed the 601 wits
+        assert!(commit_fees <= 601);
```

Comment:
A strange way to check that... I would prefer something like
```suggestion
        // t2 does not conflict with anything so it must be present
        assert!(commits_vec.contains(&t2));
        // One of t1 or t3 must be present, but not both
        assert!(commits_vec.contains(&t1) ^ commits_vec.contains(&t3));
```